doc: introduce PoseidonStore

myoungwon opened the pull request:
Introduce PoseidonStore for crimson

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>
Signed-off-by: Jiwoong Park <jw12.park@samsung.com>
Signed-off-by: Kyuho Son <kyuho.son@samsung.com>
Signed-off-by: Sungkyu Park <sk921.park@samsung.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

myoungwon commented:
@athanatos This PR is what I suggested via email. Please review if you have time to review.

athanatos commented:
I'm out this week, but I'll have a look on Monday.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
```diff
@@ -0,0 +1,254 @@
+  - Pros
+    - No host-side GC is required
+    - Fast lookup
+    - No space overhead
```
Internal fragmentation -- can be as bad as un-gc'd space.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
```diff
@@ -0,0 +1,254 @@
+    - No space overhead
+  - Cons
+    - More writes occur to record the data (metadata and data section are separated)
+    - It cannot support transaction. WAL is required to support transactions
```
It *can* support transactions -- using a WAL.  Just replace this line with "Some form of WAL required to ensure update atomicity in the general case."

athanatos commented on doc/dev/crimson/poseidonstore.rst:
we intend

athanatos commented on doc/dev/crimson/poseidonstore.rst:
SB?

tchaikov replied:
guess it stands for superblock.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I think you are confusing the ObjectStore level WAL/journal that exists in each ObjectStore implementation with the osd level PGLog.  You don't control the PGLog entry lifetime and cannot assume that there is any "checkpoint" you have access to.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
Not generally, ceph level objects/files are broken into many rados level objects.  There's a config defined max rados object size.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I think this is a *very* bad idea for a bunch of reasons:
1) In general, a particular object will be overwritten on all replicas at a similar point in time, so you'd have a decent chance of *no* replica having a valid copy of the object in the case of a correlated power loss (correlated power losses are common).
2) "we need to detect it" This is worse than it seems.
  - The WAL does give you an idea of where there were in-progress writes, but you then need to map those blocks onto rados level objects and then do a scan on those objects against other osds with the same PG.
  - This process would necessarily have to happen before we actually trust *any* of the data in any pg on an osd after restart.
  - It essentially amounts to a new incredibly complicated phase of peering (keep in mind, it's likely after a correlated crash that *all* osds in a PG have some objects that are invalid) where we have to do a sort of limited deep scrub before we can serve OSD.
  - Even without the complexity and correctness problems above, restarting a single OSD uncleanly would require a bunch of other osds to immediately do a bunch of reads (total IO proportional to max write rate of the osd which restarted) before the OSD in question can be allowed to serve IO.

I do not think you have fully thought through the implications of what using other replicas to heal inconsistent writes would entail.

athanatos commented:
I don't think the replica based recovery scheme is workable.

myoungwon commented:
Based on your feedback, we have modified the design of Poseidon Store to prevent "partial write" data corruption during power loss; that is, it doesn't rely on replica's data for recovery anymore. In the new design, for incoming writes, data is handled differently depending on the request size; data is either written twice (WAL) or written in a log-structured manner.

**(1) If Request Size ≤ Threshold** (similar to minimum allocation size in BlueStore)
Write data and metadata to [WAL] —flush—> Write them to [Data section (in-place)] and [Metadata section], respectively.

* Since the CPU becomes the bottleneck for small I/O workloads, in-place update scheme is used.
* Double write for small data may be better than host-GC overhead in terms of CPU consumption in the long run

**(2) Else if Request Size > Threshold**
Append data to [Data section (log-structure)] —> Write the corresponding metadata to [WAL] —flush—> Write the metadata to [Metadata section]

* For large I/O workloads, the host can afford process host-GC
* Also, the host can garbage collect invalid objects more effectively when their size is large

Note that Threshold can be configured to a very large number so that only the scenario (1) occurs.
With this design, we can control the overall I/O procedure with the optimizations for crimson
as described above.

Also, it seems that pglog needs to be treated like other normal data.

What do you think?

athanatos commented:
That arrangement is pretty standard and is likely workable.

In 2), given that you are already relying heavily on drive gc, I don't really see a reason for you to restrict yourself to sequential appends.  You can simply choose to write into unallocated space and then perform a journal commit which atomically marks the space as used and links it into the other metadata trees (as with bluestore and other filesystems).

Next, I'd like to see an exploration of some of the larger design choices:
- allocation
- free space tracking
- onode lookup
- omap

myoungwon commented:
First, we agree with your comment that in the case (2) writing data into unallocated space is a better idea (similar to BlueStore).
As you mentioned, we introduce some of the design choices (see below figure).

• onode lookup
Upon onode lookup requests
First, we hash CID (collection ID) to get the hash index of the hash table. Note that the entire hash table can be pre-allocated when we limit the maximum number of objects that can be created in an AG. With the hash index, the buckets are scanned until it finds the hash entry for the target object ID. 
A hash entry contains <CID, OID, Type, Offset, Segment>.
With the hash entry, we can find the disk location of the target onode.
Since the OID meta entry in the segment contains <OID, Chunk Number, (Offset, len), (Offset, len), … >, we can read the onode from the target chunk.


• allocation
Allocation Groups
Entire disk space is divided into a number of equally sized chunks called Allocation Groups (AG).
Each reactor thread in crimson-osd is responsible for a set of AGs (# of AGs / # of threads).
Each AG has its own data structures to manage the disk partition.

Data allocation
As we explained above, the entire hash table is pre-allocated in each AG. Given CID and OID, we can map any data to the segment and offset. A segment is comprised of contiguous blocks. Blocks can be allocated by searching the free space tracking data structure (we explain below).
Based on onode lookup, we probably know where OID is located via segment number. And, segment contains multiple entries such as oid meta entries (object_info_t, data etc.), small data, and data. (segment header describe what entries exists there).
So a single write can store a transaction data at once.


• free space tracking
We can use either data block bitmap in EXT4 or extent-based B+tree in XFS for free space tracking.
Our first prototype will be implemented based on the block bitmap for the sake of implementation.
Regarding allocation bitmap, we have a plan to re-use the existing the design of bitmapfreelistmanager and bitmapallocator in Bluestore (Underlying storage should be changed KV db to raw device)

• omap
In this design (see below figure), omap is not different from data and onode. They all can be retrieved via hash table lookup

• fragmentation
Internal fragmentation
We pack different type of data/metadata in a single block as many as possible to reduce internal fragmentation. 
Extent-based B+tree may help reduce this further by allocating contiguous blocks that best fit for the object

External fragmentation
Frequent object create/delete may lead to external fragmentation
In this case, we need cleaning work (GC-like) to address this.
For this, we are referring the NetApp’s Continuous Segment Cleaning, which seems similar to the SeaStore’s approach
Countering Fragmentation in an Enterprise Storage System (NetApp, ACM TOS, 2020)



          +-------------+-------------------+-------------+
           | Object meta | Allocation bitmap | Data blocks |
           +----+--------+-------------------+------+------+
                |                                   
                | CID                               
                |                                   
            +---+---+                                           
            | Hash  |                                           
            +---+---+                                           
                |                                               
                v                                               
           +---------+---------+---------+                      
           | Bucket1 | ...     | BucketN |                      
           +=========+=========+=========+                          +---------------+
           | Header  | ...     | ...     |                          | OID           |<---------------+
           +---------+---------+---------+                          | Num Chunk     |                |
        +--| Entry0  | ...     | ...     |                          | <Offset, len> |                |
        |  +---------+---------+---------+                          | <Offset, len> |                |
        |  | ...     | ...     | ...     |                          | ...           |                |
        |  +---------+---------+---------+                          +---------------+                |
        |  | EntryK  | ...     | ...     |                                                           |
        |  +---------+---------+---------+                                                           |
        |                                                                                            |
        |                                                                                            |
        |  +---------------+  +----------+  +----------+                      segment                |
        +->| object_info_t |  | omap     |  | data     |               +------------+------------+   |
           +===============+  +==========+  +==========+          +--->| Block0     | Block1     |   |
           | CID, OID      |  | CID, OID |  | CID, OID |          |    +============+============+   |
           | Type          |  | Type     |  | Type     |          |    | Header     |            |   |
           | Offset        |  | Offset   |  | Offset   |          |    +------------+            +   |
           | Segment       |  | Segment  |  | Segment  |----------+    | Small data | Data       |   |
           +---------------+  +----------+  +----------+               +------------+            +   |
                                                                    +--| OID meta Entries|       |   |
                                                                    |  +------------+------------+   |
                                                                    |                                |
                                                                    +--------------------------------+

athanatos commented:
I'll read this in more detail later this week, but I'm not sure you need to consider the collection id at all on disk.  At one point in the past, it was possible for the same oid to exist in two different collections, but that is no longer possible.  Moreover, the collection split operation is required to be basically constant time w.r.t. the number of objects in the collection.

athanatos commented:
Also, please integrate the above information into the file so I can have a look.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
As I mentioned, you don't want to include the collection ID here as you don't want to have to rehash every onode upon collection split.

myoungwon replied:
See below

athanatos commented on doc/dev/crimson/poseidonstore.rst:
How is this hash table represented on disk?  Are mutations to it included in the journal?

myoungwon replied:
See below

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I don't quite follow.  What do you mean by "probably know where OID is located via segment number"?  If you always pack all information for a write into a single segment, how is it possible for the data for two arbitrary oids to be mutated in the same transaction?  Are you simply describing the journal entry here with actual data writeout to happen lazily?  Are all transactions (on the whole PS?  on the reactor?) written to the same segment until it ends?

myoungwon replied:
> Are you simply describing the journal entry here with actual data writeout to happen lazily?

Yes.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
Same question, how do you plan on representing this on disk and how is it represented in the journal?  Is free space tracked per-AG?

athanatos replied:
This bit can be left vague for now, but I'd like some information about what you plan on doing in the short term and at least a hint at how you plan on evolving it in the future.

myoungwon replied:
See below

athanatos commented on doc/dev/crimson/poseidonstore.rst:
It is though.  Users assume that iterating an object's omap in lexicographic order is efficient -- this is a core assumption of radosgw's bucket index implementation.

myoungwon replied:
See below

athanatos replied:
You mention hash table here, did you actually mean btree?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
Are cross-AG transactions possible?  How would you plan on moving free space between reactors in the event that there are usage imbalances.  I'm fine with some vagueness here, but I'd like some design hints.

myoungwon replied:
See below

athanatos commented on doc/dev/crimson/poseidonstore.rst:
That's not the question.  Over time, data packed into your blocks will become obsolete and those blocks will contain "free" space that can't be used for allocation.  Is the plan to have a background process to reclaim under-used blocks?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I'd expect that as you implement this, the actual implementation will change dramatically.  I'm more interested in generally how you structure the cache w.r.t. how other components interact with it and how lifetimes are managed.

myoungwon replied:
See below

athanatos commented on doc/dev/crimson/poseidonstore.rst:
3. here is an incompletely written WAL entry that you can detect with a checksum, right?

myoungwon replied:
Hm, do we need checksum here? because we can write a large data unit atomically by using Atomic Write Unit Normal.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I don't really understand this portion.

myoungwon replied:
Removed

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I think we agreed not to use replicas log replay/startup?

myoungwon replied:
Removed.

athanatos commented:
I left some questions I have on the design.  As I mentioned above, I'm fine with some vagueness, but I'd like some more hints about your strategy.  Looks to me like there are some incomplete sections at the end of the document -- did you mean to add more details there before I continue reviewing?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I'd like some kind of information at least about
1. Is there a single WAL for the whole store?  Does each AG get a WAL?  Are cross-WAL transactions possible?
2. What information goes into the WAL? (metadata updates, free space updates, data)?
3. How does the space consumed by the WAL interact with the free space tracking metadata?
...

myoungwon commented:
We've slightly improved proposed design based on your feedback.


Flat namespace
Object data, object meta, omap, and xattrs are all the data related to an object (we call it a head object from ghobject).
PoseidonStore adds different suffixes to the head object ID to create PSD_OID to distinguish them 
(e.g., Assuming head object ID = 12.e4 -> Object data's PSD_OID = 12.e4_data, Object meta's PSD_OID = 12.e4_meta, omap's PSD_OID = 12.e4_omap_key1, xattr's PSD_OID = 12.e4_xattr_key1).
By eliminating suffixes, we can recover their head object ID. Thus, we can know they are all related to the same object.
Note that because all objects are sorted by their PSD_OIDs in lexicographic order, xattrs and omap can be iterated in lexicographic order.
For object deletion, all the objects having the same head object ID are searched and deleted.

Lookup 
As described in the figure, we replaced hash table on Object meta with b+tree to find out items efficiently.
This is because we can control the tree based on the policy we define and it can do fast lookup.
There is no on-disk onode data structure. To fill the in-memory onode data structure, object meta, xattrs, 
and omap of the target object can be retrieved using the respective PSD_OIDs.
Also, we use PSD_OID as a key to find the leaf node that contains extent.
Data type in leaf node indicates if the pointing data is inline data or an extent array of <start block num, block count>.
According to the data type, it reads the blocks needed for the object.
If you look at the figure as below, there is the leafnode which contains extent info. Using 
extent info, we can know where the extent is located. Extent contains where data 
chunks locate by using extent map, so we finally figure out the data location.

Data allocation + Freelist management
Just to clarify, we use the term "extent" instead of "segment" that we used in the previous version of doc.
The freespace is tracked on a per-AG basis.
The initial version of PoseidonStore will use bitmaps for the validity of all blocks in the AG. 
The bitmaps are pre-allocated in the allocation bitmap partition.
Upon new object write, we allocate an extent of contiguous blocks large enough to fit data by referring to the allocation bitmaps.
We may use two extent-based b+trees for efficient contiguous free space tracking; one by block number and another by block count (similar to XFS).
We leave this for future work. The means of free space tracking can be configured at the initial disk setup (cannot be changed after configuration).

Durablity of Object Meta 
whenever the tree is updated, the modification is first append at the WAL. Then, flush it
to the data partition.

Cache
There are mainly two cache data structures; onode cache and block cache.
Since PoseidonStore does not have the on-disk onode strucutre, there needs to be in-memory onode structure for fast tracking of the objects.
It looks like below.
Onode {
    metadata
    xattr map <key, value>
    omap <key, value>
    extent_map
}
To fill the in-memory onode data structure, object meta, xattrs, and omap of the target object can be retrieved using the respective PSD_OIDs.
Block cache is used for caching a block contents and supporting transactions.
For a transaction, all the updates to blocks (including object meta block, allocation bitmap block, data block) are first performed in the in-memory block cache.
After writing a transaction to the WAL, the dirty blocks are flushed to their respective locations in the respective partitions.
PoseidonStore can configure cache size for each type. Simple LRU cache eviction strategy can be used for both.




WAL
Each AG has a WAL.
The datas written to the WAL are all metadta updates, free space update and data.
Note that only data smaller than the predefined threshold needs to be written to the WAL.
The larger data is written to the unallocated free space and its onode's extent_map is updated accordingly (also on-disk extent map).
We statically allocate WAL partition aside from data partition pre-configured.

Allocation Group (with cross-AG transaction)
Entire disk space is divided into a number of equally sized chunks called Allocation Groups (AG).
Each reactor thread in crimson-osd is responsible for a set of AGs (# of AGs / # of threads). 
Each AG has its own data structures to manage the disk partition.
A collection is stored into an AG, not across AGs.
The prefixes of the parent collection ID (original collection ID before collection splitting. That is, hobject.hash) is hashed to map any collections to AGs allocated for the reactor thread.
We can use BlueStore's approach for collection splitting, changing the number of significant bits for the collection prefixes.
Because the prefixes of the parent collection ID do not change even after collection splitting, the mapping between the collection and AG is maintained.
The number of AGs may be configured to match the number of CPUs allocated for each disk so that each AG can hold a number of objects large enough for cross-AG transaction not to occur.
In case of need of cross-AG transaction, we could simply use the per-AG lock (acquire the source and target locks before processing the cross-AG transaction).
For the load unbalanced situation, we adjust the mapping between collection and AG and the mapping between reactor thread and collection. We need more details.


      +-------------+-------------------+-------------+
       | Object meta | Allocation bitmap | Data blocks | --------------------------------
       +----+--------+-------------------+------+------+               |                |
            |                                                          |                | 
            | OID                                                      |                | 
            |                                                          |                |
        +---+---+                                                      |                |
        | Root  |                                                      |                |
        +---+---+                                                      |                |
            |                                                          |                |
            v                 Tree                                     |                |
       +---------+---------+---------+                                 |                |
       | Subtree | ...     | Subtree |                                 v                |
       +=========+=========+=========+                          +---------------+       |
       | Header  | ...     | ...     |                          | OID           |<-----------------+
       +---------+---------+---------+                          | Num Chunk     |       |          |
    +--| Entry0  | ...     | ...     |                          | <Offset, len> |       |          |
    |  +---------+---------+---------+                          | <Offset, len> |-------|----+     |
    |  | ...     | ...     | ...     |                          | ...           |       |    |     |
    |  +---------+---------+---------+                          +---------------+       |    |     |
    |  | EntryK  | ...     | ...     |                                                  |    |     |
    |  +---------+---------+---------+                                                  |    |     |
    |                                                                                   |    |     |
    |                                                                                   |    |     |
    |  +---------------+  +----------+  +-------------+                                 v    v     |
    +->| leafnode      |  | leafnode |  | leafnode    |               +------------+------------+  |
       +===============+  +==========+  +=============+               | Block0     | Block1     |  |
       | OID           |  | OID      |  | OID         |               +============+============+  |
       | Block no      |  | Block no |  | Block no    |               | Free       |            |  |
       | Offset        |  | Offset   |  | Offset      |               +------------+            +  |
       | Extent        |  | Extent   |  | Extent      |----------+    | Small data | Data       |  |
       +---------------+  +----------+  +-------------+          |    +------------+            +  |
                                                                 +-+->| Extent     |            |  |
                                                                      +------------+------------+  |
                                                                          |                        |
                                                                          +------------------------+

myoungwon commented:
@athanatos Can you review this if you have time to take a look?

athanatos commented:
Is all of the above reflected in the doc?

myoungwon commented:
Updated.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
Are you specifically excluding rgw?

athanatos replied:
What about omap?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
How do you do this atomically without including it in the WAL write?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
You don't need to write out the collection, as we discussed previously.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I don't understand what this means.  For a large write, I'd assume that you need to write and persist the data portion *prior* to the WAL metadata update (as bluestore does).

athanatos replied:
Wait, is the idea that for a large write you'd do *three* writes?  A WAL (tx begin) ->data write->WAL (tx end)?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
"whenever needed".  That would be every time you want to signal that a transaction has committed, right?  You could batch transactions and reduce the number of flushes, but only at the cost of delaying completion on the transactions within the batch.

athanatos replied:
Actually, I'm unsure of what you mean by flush here.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
"flush happens" I don't understand.  Do you mean that you don't have enough preallocated usable space to do the data write?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
You are proposing a btree design, but you are not accounting for the overhead of the additional updates to the rest of the btree.  Note that temporal locality, cache size, and tolerable wal replay time may reduce the number of effective writes.  I think I'd like to see some detail on that.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I've seen "flush" used throughout this document, but I'm not sure what you mean.  Do you mean writing out dirty metadata/data from the cache to its final location in order to free up space in the wal, or do you mean something else?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
So your metadata btree stores arbitrary length strings sorted lexicographically?

athanatos replied:
How do you plan on addressing COW clones?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
This has a disadvantage: scanning objects for backfill and scrub will require enumerating omap/xattr data.  How do you plan on representing an object with no xattrs or omap?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
all the keys -- obviously a single object deletion only removes a single object.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
"replaced hash table" what hash table?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
How would omap be involved in the onode?

There's a really really good reason not to do it this way: onode lookup (that is, checking for object existence and size) is *by far* the most common read operation -- requiring that the onode be reconsituted from several keys means you need to spend a bunch of cpu cycles doing so.  This alone is one of the largest problems with bluestore as it exists today.  You probably want to arrange for the binary representation of the onode on disk to be contiguous and interpretable directly.  Eventually, we'll want most of the object_info_t to work this way as well, though that will require refactoring elsewhere.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
At what granularity?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
Would this replace the bitmap?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I haven't seen anything about a hash table.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
"Each AG has a WAL".  This to mean implies that you cannot have a transaction that spans two AGs.  I think this is going to be a problem for you.  Your metadata structures and data structures pertaining to a particular object may be allocated over multiple AGs -- how would you perform an atomic update under those conditions?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I *think* you mean here that each reactor thread gets a static set of AGs.  You've gone to the trouble of introducing crimson reactor threads and AGs in this document, please connect those concepts to "partition" here.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I asked above, but is the idea that you don't care about the ondisk onode representation *because* you assume you'll be able to cache all of them?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
You definitely cannot store all of the omap keys in memory as part of the onode.  I've seen cases where an omap size for a single rgw bucket index object exceeds 80GB.  Is your assumption that you won't use this implementation for osds which store objects with more than 1-5 omap keys used exclusively for extremely small omap data?  If you choose that route, you'll be locking yourself out of most cephfs and rgw configurations.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
FYI: this is explicitly ok -- existing usage assumes that writes will be available to read prior to commit.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I need more detail on this part, I think you are missing a lot here that will matter hugely -- no cross ag transactions means every PG is local to a single AG and you cannot move less than a full PG between AGs.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
Ok, but does this mean that your WAL transaction design needs special support for cross-WAL transactions?  Certainly, the possibility means that operations in different WALs are at least partially orderable -- how do you plan on dealing with that?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I don't understand this table.  What is a Subtree here?

athanatos commented:
When you address the above comments, please revise the existing text to ensure that you don't leave anything out of date, and please integrate any new information into the correct sections in a manner that makes your intent as clear and concise as possible.

myoungwon commented:
@athanatos We've update the document. Please take a look. As for cross-partition transaction (between PGs), 
can you let me know the case of cross transaction? Because there is no data movement when pg split happens.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
Metadata

myoungwon replied:
Fixed.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
Does this mean "Flush data to trim WAL as necessary"?

myoungwon replied:
That is what I was saying.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I don't understand -- the oid would have to be the hobject_t from the osd and ultimately from the client.  Where would those bits come from?

myoungwon replied:
A few bits (leftmost bits of the hash) of the OID determine a sharded partition where the object is located.
  For example, if the number of partitions is configured as four, The entire space of the hash in hobject_t
  can be divided into four domains (0x0xxx ~ 0x3xxx, 0x4xxx ~ 0x7xxx, 0x8xxx ~ 0xBxxx and 0xCxxx ~ 0xFxxx).

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I don't understand this part.  Can you discuss how this maps onto the clone() and clone_range() operations?

myoungwon replied:
See below

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I need more information here.

myoungwon replied:
As for CoW/Clone, a clone has an onode like other normal objects.
As shown the figure as below, by using shared_blob tracked by shared_blob_list in the superblock,
the blocks for the clone are managed using shared_block with reference count.


    +---------+        +---------+
    | Onode A |---     | Clone A |
    +---------+  |     +---------+
        v       ---v      v
    +--------------+----------------------------+
    | Data block 1 | Data block 2 (shared_blob) |
    +--------------+----------------------------+

athanatos replied:
I don't really understand what the clones element of an onode means.  By "shared_blob_list" in the superblock, what do you mean?

myoungwon replied:
Hm... Isn't this similar to existing Ceph or btrfs?
Assuming that there is an object A,
clones from the object A will have an independent OID like xxx:0 and xxx:1 (unlike xxx:head).
So, these clones will have its own onode.

Although each clone has its own onode, data blocks should be shared between the original object and clones
if there are no changes on them to minimize the space overhead.
To do so, the reference count for the data blocks is needed to manage those shared data blocks.

To deal with the data blocks which has the reference count, the practical way is to use the shared block information recorded in the global area (within a sharded partition) like superblock. 
shared_blob_list means shared block information for the data blocks.

athanatos replied:
I guess I don't understand specifically why each onode needs to in some sense track clones.  I expect that some extent indirections necessarily point at an indirection capable of handling ref counting (the shared_blob structure), but I don't understand yet how the shard_blob works or what the onode 'clones' member does.

myoungwon replied:
I removed 'clones' in the onode, there is no need to have such member variable. 
Like current Ceph, Extent has a Blob, and the Blob can have shared_blob if any.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
Why does the global section have an onode B+ tree or data blocks?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
You seem to suggest that this will be a radix tree or a B+ tree in different places, which is it?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
A radix tree is a pretty unusual choice for omap data, can you expand on why?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
Why is the tree information embedded in the onode?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I think you'd be better off caching omap via a more general block cache.  Explicitly representing each key seems complicated and cpu expensive.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
This diagram is pretty confusing.  There is no distinction between objects and clones, so I think you mean object A and object B.  Second, where is the refcount information for the shared_block stored relative to object A's onode, object B's onode and the actual data blocks?  I think that would be the most important thing to convey.

athanatos commented:
I think this is almost ready to merge.  I think the last thing I'd like to see (other than my comments above) is a list of implementation stages/steps outlining how you plan on breaking the overall project into steps/PRs.

myoungwon commented:
I addressed your comments. Can you take a look? In addition, I added a section about the plans.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
I continue to be confused by this part.  I assume that the point of an onode radix tree is to index onodes by ghobject_t.  From the fact that you have 'onode childs' and 'onode parent_node' I guess that the internal nodes of your tree for some reason contain onodes.  Why?  Especially with a radix tree, one generally chooses to have only keys in internal nodes to reduce fan-out with the actual values (onodes, here) only in leaf nodes.

Also, where do you plan on storing the xattrs?

myoungwon replied:
Sorry for confusing. Onode here is an onode pointer to indicate another onode. Like xa_node in linux kernel, parent is an 
onode pointer to root, and childs is a pointer array the onode belongs to.

I think the omap structure is able to handle the xattrs because they have the same data model where each key is associated with the value. Do I need a separated data structure for xattrs?

athanatos replied:
I think you are describing a basic in-memory n-ary tree implementation.  A radix tree as you insist that you are using here is quite different, particularly when implemented on disk.  For the purposes of this discussion, you probably want to remove references to the indexing structures from the onode itself -- they are orthogonal (or at least mostly so).

You stated that the onode is fixed size.  I suggest that you store the xattrs within the onode and permit it to be variably sized -- if only because the ghobject key already is.

myoungwon replied:
I got it.

athanatos commented on doc/dev/crimson/poseidonstore.rst:
There is confusion above about what a radix tree is and how you mean to use it.  For now, add a planning item here:

```First, submit a PR against this file with a more detailed on disk layout and lookup strategy for the onode radix tree.

Follow up with implementations based on the above design once design PR is merged.```

I'm also not sure what you mean below by "onode can interact between onodes".

myoungwon replied:
Updated.

myoungwon commented:
I've updated the doc file based on your comment. Can you take a look?

athanatos commented on doc/dev/crimson/poseidonstore.rst:
xattrs here should simply be inline, I really don't think you want to pay additional lookups for it.  Are you aware of how the osd code (crimson as well at the moment) uses xattrs?

myoungwon replied:
Xattr intends to store the small data which can embed in the onode or store EC data . Otherwise, omap is typically larger data than xattr. Right? My intention is that additional space is necessary if inline xattrs become full. Anyway, fixed it.

athanatos replied:
The osd uses an xattr to store the object_info_t on every object and a second one to store the snapset on the head object.  These tend to be accessed on every read and write.  It would be unfortunate to have to do additional reads to access them.

athanatos replied:
I suppose using a btree for the xattrs is fine as long as there's an optimization where they are internal to the onode if they don't take up too much space.

myoungwon replied:
I think inline xattr is better.
If xattr is large or the number of xattr  entries grows, the xattr can be moved into the b+tree, omap the onode already has.

myoungwon replied:
Any other comments?
